domain/machine/state/state.go

```diff
@@ -569,3 +645,70 @@ func (st *State) IsMachineRebootRequired(ctx context.Context, uuid string) (bool
 
 	return isRebootRequired, errors.Annotatef(err, "requiring reboot of machine %q", uuid)
 }
+
+// GetMachineParentUUID returns the parent UUID of the specified machine.
+// It returns a NotFound if the machine does not exist.
+// It returns a MachineHasNoParent if the machine has no parent.
+func (st *State) GetMachineParentUUID(ctx context.Context, mName machine.Name) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	// Prepare query for machine UUID.
+	machineNameParam := machineName{Name: mName}
+	machineUUIDoutput := machineUUID{}
+	query := `SELECT uuid AS &machineUUID.* FROM machine WHERE name = $machineName.name`
+	queryStmt, err := st.Prepare(query, machineNameParam, machineUUIDoutput)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	// Prepare query for parent UUID.
+	parentUUID := ""
+	parentUUIDParam := machineParent{}
+	parentQuery := `
+SELECT parent_uuid AS &machineParent.parent_uuid
+FROM machine_parent WHERE machine_uuid = $machineUUID.uuid`
+	parentQueryStmt, err := st.Prepare(parentQuery, machineUUIDoutput, parentUUIDParam)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Query for the machine UUID.
+		err := tx.Query(ctx, queryStmt, machineNameParam).Get(&machineUUIDoutput)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.NotFoundf("machine %q", mName)
+			}
+			return errors.Annotatef(err, "querying UUID for machine %q", mName)
+		}
+
+		// Query for the parent UUID.
+		err = tx.Query(ctx, parentQueryStmt, machineUUIDoutput).Get(&parentUUIDParam)
+		if err != nil {
+			if errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Annotatef(machineerrors.MachineHasNoParent, "machine %q", mName)
+			}
+			return errors.Annotatef(err, "querying parent UUID for machine %q", mName)
+		}
+
+		parentUUID = parentUUIDParam.ParentUUID
+
+		// Protect against a grandparent
+		machineUUIDoutput.UUID = parentUUID
+		err = tx.Query(ctx, parentQueryStmt, machineUUIDoutput).Get(&parentUUIDParam)
+		// No error means we found a grandparent.
+		if err == nil {
+			return errors.Annotatef(machineerrors.GrandParentNotAllowed, "machine %q", mName)
+		} else if !errors.Is(err, sqlair.ErrNoRows) {
```

Comment:
**todo:** `else` not required here, since previous if-branch is a return statement. 

Usually, in golang, we use early return style with no else. I think we should stick on it (and it may broke picky static analysis some day)